File: OximyMac/Services/MDMConfigService.swift
import Foundation

/// Service to detect and read MDM-managed configuration profiles.
///
/// MDM solutions (Jamf, Kandji, Intune, etc.) can deploy configuration profiles
/// that set managed preferences in the app's UserDefaults domain (com.oximy.mac).
/// This service reads those preferences and applies MDM policies.
///
/// Configuration profile keys:
/// - ManagedDeviceToken: Pre-provisioned device API token
/// - ManagedDeviceId: Pre-assigned device identifier
/// - ManagedWorkspaceId: Organization/workspace ID
/// - ManagedWorkspaceName: Display name for workspace
/// - ManagedSetupComplete: Skip all setup UI (Bool)
/// - ManagedEnrollmentComplete: Skip enrollment UI only (Bool)
/// - ManagedCACertInstalled: CA cert deployed via MDM (Bool)
/// - ForceAutoStart: Prevent user from disabling auto-start (Bool)
/// - DisableUserLogout: Hide logout option in UI (Bool)
/// - DisableQuit: Prevent CMD+Q termination (Bool)
/// - APIEndpoint: Custom API URL (String)
@MainActor
final class MDMConfigService: ObservableObject {
    static let shared = MDMConfigService()

    // MARK: - Constants

    /// The app's bundle identifier for reading managed preferences
    private let appIdentifier = "com.oximy.mac" as CFString

    // MARK: - Managed Preference Keys

    /// Keys for MDM-managed preferences (set via configuration profile)
    enum ManagedKey: String {
        // Credentials
        case deviceToken = "ManagedDeviceToken"
        case deviceId = "ManagedDeviceId"
        case workspaceId = "ManagedWorkspaceId"
        case workspaceName = "ManagedWorkspaceName"

        // Setup bypass
        case setupComplete = "ManagedSetupComplete"
        case enrollmentComplete = "ManagedEnrollmentComplete"
        case caCertInstalled = "ManagedCACertInstalled"

        // Lockdown controls
        case forceAutoStart = "ForceAutoStart"
        case disableUserLogout = "DisableUserLogout"
        case disableQuit = "DisableQuit"

        // Configuration
        case apiEndpoint = "APIEndpoint"
        case heartbeatInterval = "HeartbeatInterval"
    }

    // MARK: - Published State

    /// Whether this device is managed by an MDM
    @Published private(set) var isManagedDevice: Bool = false

    /// Timestamp of last config check
    @Published private(set) var lastConfigCheck: Date?

    // MARK: - Initialization

    private init() {
        checkManagedStatus()
    }

    // MARK: - Public Methods

    /// Check if any managed preferences are set and apply configuration
    func checkManagedStatus() {
        // Use CFPreferences to read managed preferences (from /Library/Managed Preferences/)
        // This is required because UserDefaults.standard doesn't automatically read MDM profiles
        let hasToken = getManagedString(for: .deviceToken) != nil
        let hasSetupFlag = getManagedValue(for: .setupComplete) != nil
        let hasEnrollmentFlag = getManagedValue(for: .enrollmentComplete) != nil
        let hasForceAutoStart = getManagedValue(for: .forceAutoStart) != nil

        isManagedDevice = hasToken || hasSetupFlag || hasEnrollmentFlag || hasForceAutoStart
        lastConfigCheck = Date()

        if isManagedDevice {
            print("[MDMConfigService] Managed device detected")
            logManagedConfig()
        } else {
            print("[MDMConfigService] Not a managed device")
        }
    }

    // MARK: - Core Foundation Preference Reading

    /// Read a managed preference value using CFPreferences (reads from /Library/Managed Preferences/)
    private func getManagedValue(for key: ManagedKey) -> Any? {
        let value = CFPreferencesCopyAppValue(key.rawValue as CFString, appIdentifier)
        return value as Any?
    }

    /// Read a managed string preference
    private func getManagedString(for key: ManagedKey) -> String? {
        guard let value = CFPreferencesCopyAppValue(key.rawValue as CFString, appIdentifier) else {
            return nil
        }
        return value as? String
    }

    /// Read a managed boolean preference
    private func getManagedBool(for key: ManagedKey) -> Bool? {
        guard let value = CFPreferencesCopyAppValue(key.rawValue as CFString, appIdentifier) else {
            return nil
        }
        // Handle both Boolean and Number types
        if let boolValue = value as? Bool {
            return boolValue
        }
        if let numValue = value as? NSNumber {
            return numValue.boolValue
        }
        return nil
    }

    /// Read a managed integer preference
    private func getManagedInt(for key: ManagedKey) -> Int? {
        guard let value = CFPreferencesCopyAppValue(key.rawValue as CFString, appIdentifier) else {
            return nil
        }
        if let intValue = value as? Int {
            return intValue
        }
        if let numValue = value as? NSNumber {
            return numValue.intValue
        }
        return nil
    }

    /// Apply managed configuration to app's standard UserDefaults
    /// Call this during app initialization to sync MDM settings
    func applyManagedConfiguration() {
        guard isManagedDevice else { return }

        let defaults = UserDefaults.standard

        // Sync device credentials
        if let token = managedDeviceToken {
            defaults.set(token, forKey: Constants.Defaults.deviceToken)
            print("[MDMConfigService] Applied managed device token")
        }

        if let deviceId = managedDeviceId {
            defaults.set(deviceId, forKey: Constants.Defaults.deviceId)
        }

        if let workspaceId = managedWorkspaceId {
            defaults.set(workspaceId, forKey: Constants.Defaults.workspaceId)
        }

        if let workspaceName = managedWorkspaceName {
            defaults.set(workspaceName, forKey: Constants.Defaults.workspaceName)
        }

        // Sync setup state
        if managedSetupComplete {
            defaults.set(true, forKey: Constants.Defaults.setupComplete)
            print("[MDMConfigService] Applied managed setup complete flag")
        }

        // Sync API configuration
        if let apiEndpoint = managedAPIEndpoint {
            defaults.set(apiEndpoint, forKey: Constants.Defaults.apiEndpoint)
        }

        if let interval = managedHeartbeatInterval {
            defaults.set(interval, forKey: Constants.Defaults.heartbeatInterval)
        }
    }

    // MARK: - Credential Accessors (with 3-tier fallback)

    /// Pre-provisioned device token from MDM
    /// Priority: MDM > remote-state (API) > default (nil)
    var managedDeviceToken: String? {
        if let mdmValue = getManagedString(for: .deviceToken) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.managedDeviceToken { return apiValue }
        return nil
    }

    /// Pre-assigned device ID from MDM
    /// Priority: MDM > remote-state (API) > default (nil)
    var managedDeviceId: String? {
        if let mdmValue = getManagedString(for: .deviceId) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.managedDeviceId { return apiValue }
        return nil
    }

    /// Organization workspace ID from MDM
    /// Priority: MDM > remote-state (API) > default (nil)
    var managedWorkspaceId: String? {
        if let mdmValue = getManagedString(for: .workspaceId) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.managedWorkspaceId { return apiValue }
        return nil
    }

    /// Display name for workspace from MDM
    /// Priority: MDM > remote-state (API) > default (nil)
    var managedWorkspaceName: String? {
        if let mdmValue = getManagedString(for: .workspaceName) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.managedWorkspaceName { return apiValue }
        return nil
    }

    // MARK: - Setup Bypass Accessors (with 3-tier fallback)

    /// Skip all setup UI when true (requires device token)
    /// Priority: MDM > remote-state (API) > default
    var managedSetupComplete: Bool {
        if let mdmValue = getManagedBool(for: .setupComplete) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.managedSetupComplete { return apiValue }
        return false
    }

    /// Skip enrollment UI only when true (requires device token)
    /// Priority: MDM > remote-state (API) > default
    var managedEnrollmentComplete: Bool {
        if let mdmValue = getManagedBool(for: .enrollmentComplete) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.managedEnrollmentComplete { return apiValue }
        return false
    }

    /// CA certificate is already installed via MDM profile
    /// Priority: MDM > remote-state (API) > default
    var managedCACertInstalled: Bool {
        if let mdmValue = getManagedBool(for: .caCertInstalled) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.managedCACertInstalled { return apiValue }
        return false
    }

    // MARK: - Lockdown Control Accessors (with 3-tier fallback)

    /// Prevent user from disabling auto-start
    /// Priority: MDM > remote-state (API) > default
    var forceAutoStart: Bool {
        if let mdmValue = getManagedBool(for: .forceAutoStart) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.forceAutoStart { return apiValue }
        return false
    }

    /// Hide logout option in UI
    /// Priority: MDM > remote-state (API) > default
    var disableUserLogout: Bool {
        if let mdmValue = getManagedBool(for: .disableUserLogout) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.disableUserLogout { return apiValue }
        return false
    }

    /// Prevent CMD+Q from quitting the app
    /// Priority: MDM > remote-state (API) > default
    var disableQuit: Bool {
        if let mdmValue = getManagedBool(for: .disableQuit) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.disableQuit { return apiValue }
        return false
    }

    // MARK: - Configuration Accessors (with 3-tier fallback)

    /// Custom API endpoint from MDM
    /// Priority: MDM > remote-state (API) > default (nil)
    var managedAPIEndpoint: String? {
        if let mdmValue = getManagedString(for: .apiEndpoint) { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.apiEndpoint { return apiValue }
        return nil
    }

    /// Heartbeat interval override from MDM (in seconds)
    /// Priority: MDM > remote-state (API) > default (nil)
    var managedHeartbeatInterval: Int? {
        if let mdmValue = getManagedInt(for: .heartbeatInterval), mdmValue > 0 { return mdmValue }
        if let apiValue = RemoteStateService.shared.appConfig?.heartbeatInterval, apiValue > 0 { return apiValue }
        return nil
    }

    // MARK: - Helper Methods

    /// Whether the app should skip the enrollment phase
    var shouldSkipEnrollment: Bool {
        guard isManagedDevice else { return false }

        // Skip if we have a device token AND enrollment is marked complete
        return managedDeviceToken != nil && (managedEnrollmentComplete || managedSetupComplete)
    }

    /// Whether the app should skip all setup (enrollment + cert + proxy)
    var shouldSkipAllSetup: Bool {
        guard isManagedDevice else { return false }

        // Skip all setup if MDM says so AND we have credentials
        return managedDeviceToken != nil && managedSetupComplete
    }

    /// Log current managed configuration for debugging
    private func logManagedConfig() {
        print("[MDMConfigService] Configuration:")
        print("  - Device Token: \(managedDeviceToken != nil ? "***" : "not set")")
        print("  - Workspace Name: \(managedWorkspaceName ?? "not set")")
        print("  - Setup Complete: \(managedSetupComplete)")
        print("  - Enrollment Complete: \(managedEnrollmentComplete)")
        print("  - CA Cert Installed: \(managedCACertInstalled)")
        print("  - Force Auto-Start: \(forceAutoStart)")
        print("  - Disable Logout: \(disableUserLogout)")
        print("  - Disable Quit: \(disableQuit)")
        print("  - API Endpoint: \(managedAPIEndpoint ?? "default")")
    }
}

// MARK: - MDM Error Types

enum MDMConfigError: LocalizedError {
    case managedByMDM(String)

    var errorDescription: String? {
        switch self {
        case .managedByMDM(let action):
            return "Cannot \(action): This device is managed by your organization."
        }
    }
}

File: OximyWindows/src/OximyWindows/Services/MDMConfigService.cs
using System.ComponentModel;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using Microsoft.Win32;

namespace OximyWindows.Services;

/// <summary>
/// Service to detect and read MDM-managed configuration from Windows Registry.
/// MDM solutions (Intune, SCCM, etc.) can deploy configuration via Group Policy
/// or custom OMA-URI settings to HKLM\SOFTWARE\Policies\Oximy\.
///
/// This mirrors the macOS MDMConfigService which reads from /Library/Managed Preferences/.
/// </summary>
public class MDMConfigService : INotifyPropertyChanged
{
    private static MDMConfigService? _instance;
    public static MDMConfigService Instance => _instance ??= new MDMConfigService();

    private const string PolicyKeyPath = @"SOFTWARE\Policies\Oximy";

    // Managed Preference Keys (matching macOS names for consistency)
    private const string KeyDeviceToken = "ManagedDeviceToken";
    private const string KeyDeviceId = "ManagedDeviceId";
    private const string KeyWorkspaceId = "ManagedWorkspaceId";
    private const string KeyWorkspaceName = "ManagedWorkspaceName";
    private const string KeySetupComplete = "ManagedSetupComplete";
    private const string KeyEnrollmentComplete = "ManagedEnrollmentComplete";
    private const string KeyCACertInstalled = "ManagedCACertInstalled";
    private const string KeyForceAutoStart = "ForceAutoStart";
    private const string KeyDisableUserLogout = "DisableUserLogout";
    private const string KeyDisableQuit = "DisableQuit";
    private const string KeyAPIEndpoint = "APIEndpoint";
    private const string KeyHeartbeatInterval = "HeartbeatInterval";

    private bool _isManagedDevice;
    /// <summary>
    /// Whether this device has MDM configuration deployed.
    /// </summary>
    public bool IsManagedDevice
    {
        get => _isManagedDevice;
        private set => SetProperty(ref _isManagedDevice, value);
    }

    private DateTime? _lastConfigCheck;
    /// <summary>
    /// When the MDM configuration was last checked.
    /// </summary>
    public DateTime? LastConfigCheck
    {
        get => _lastConfigCheck;
        private set => SetProperty(ref _lastConfigCheck, value);
    }

    // Cached configuration values
    private string? _managedDeviceToken;
    private string? _managedDeviceId;
    private string? _managedWorkspaceId;
    private string? _managedWorkspaceName;
    private bool _managedSetupComplete;
    private bool _managedEnrollmentComplete;
    private bool _managedCACertInstalled;
    private bool _forceAutoStart;
    private bool _disableUserLogout;
    private bool _disableQuit;
    private string? _apiEndpoint;
    private int? _heartbeatInterval;

    // MARK: - Credential Accessors (with 3-tier fallback: MDM > API > default)

    /// <summary>
    /// Pre-provisioned device API token.
    /// Priority: MDM > remote-state (API) > default (null)
    /// </summary>
    public string? ManagedDeviceToken
    {
        get
        {
            if (!string.IsNullOrEmpty(_managedDeviceToken)) return _managedDeviceToken;
            if (!string.IsNullOrEmpty(RemoteStateService.Instance.AppConfig?.ManagedDeviceToken)) return RemoteStateService.Instance.AppConfig.ManagedDeviceToken;
            return null;
        }
    }

    /// <summary>
    /// Pre-assigned device identifier.
    /// Priority: MDM > remote-state (API) > default (null)
    /// </summary>
    public string? ManagedDeviceId
    {
        get
        {
            if (!string.IsNullOrEmpty(_managedDeviceId)) return _managedDeviceId;
            if (!string.IsNullOrEmpty(RemoteStateService.Instance.AppConfig?.ManagedDeviceId)) return RemoteStateService.Instance.AppConfig.ManagedDeviceId;
            return null;
        }
    }

    /// <summary>
    /// Organization workspace ID.
    /// Priority: MDM > remote-state (API) > default (null)
    /// </summary>
    public string? ManagedWorkspaceId
    {
        get
        {
            if (!string.IsNullOrEmpty(_managedWorkspaceId)) return _managedWorkspaceId;
            if (!string.IsNullOrEmpty(RemoteStateService.Instance.AppConfig?.ManagedWorkspaceId)) return RemoteStateService.Instance.AppConfig.ManagedWorkspaceId;
            return null;
        }
    }

    /// <summary>
    /// Display name for workspace.
    /// Priority: MDM > remote-state (API) > default (null)
    /// </summary>
    public string? ManagedWorkspaceName
    {
        get
        {
            if (!string.IsNullOrEmpty(_managedWorkspaceName)) return _managedWorkspaceName;
            if (!string.IsNullOrEmpty(RemoteStateService.Instance.AppConfig?.ManagedWorkspaceName)) return RemoteStateService.Instance.AppConfig.ManagedWorkspaceName;
            return null;
        }
    }

    // MARK: - Setup Bypass Accessors (with 3-tier fallback)

    /// <summary>
    /// Skip all setup UI (enrollment + certificate + proxy).
    /// Priority: MDM > remote-state (API) > default (false)
    /// </summary>
    public bool ManagedSetupComplete
    {
        get
        {
            if (_isManagedDevice && _managedSetupComplete) return true;
            if (RemoteStateService.Instance.AppConfig?.ManagedSetupComplete == true) return true;
            return false;
        }
    }

    /// <summary>
    /// Skip enrollment UI only.
    /// Priority: MDM > remote-state (API) > default (false)
    /// </summary>
    public bool ManagedEnrollmentComplete
    {
        get
        {
            if (_isManagedDevice && _managedEnrollmentComplete) return true;
            if (RemoteStateService.Instance.AppConfig?.ManagedEnrollmentComplete == true) return true;
            return false;
        }
    }

    /// <summary>
    /// CA certificate was deployed via MDM.
    /// Priority: MDM > remote-state (API) > default (false)
    /// </summary>
    public bool ManagedCACertInstalled
    {
        get
        {
            if (_isManagedDevice && _managedCACertInstalled) return true;
            if (RemoteStateService.Instance.AppConfig?.ManagedCACertInstalled == true) return true;
            return false;
        }
    }

    // MARK: - Lockdown Control Accessors (with 3-tier fallback)

    /// <summary>
    /// Prevent user from disabling auto-start.
    /// Priority: MDM > remote-state (API) > default (false)
    /// </summary>
    public bool ForceAutoStart
    {
        get
        {
            if (_isManagedDevice && _forceAutoStart) return true;
            if (RemoteStateService.Instance.AppConfig?.ForceAutoStart == true) return true;
            return false;
        }
    }

    /// <summary>
    /// Hide logout option in UI.
    /// Priority: MDM > remote-state (API) > default (false)
    /// </summary>
    public bool DisableUserLogout
    {
        get
        {
            if (_isManagedDevice && _disableUserLogout) return true;
            if (RemoteStateService.Instance.AppConfig?.DisableUserLogout == true) return true;
            return false;
        }
    }

    /// <summary>
    /// Prevent app termination.
    /// Priority: MDM > remote-state (API) > default (false)
    /// </summary>
    public bool DisableQuit
    {
        get
        {
            if (_isManagedDevice && _disableQuit) return true;
            if (RemoteStateService.Instance.AppConfig?.DisableQuit == true) return true;
            return false;
        }
    }

    // MARK: - Configuration Accessors (with 3-tier fallback)

    /// <summary>
    /// Custom API URL override.
    /// Priority: MDM > remote-state (API) > default (null)
    /// </summary>
    public string? APIEndpoint
    {
        get
        {
            if (!string.IsNullOrEmpty(_apiEndpoint)) return _apiEndpoint;
            if (!string.IsNullOrEmpty(RemoteStateService.Instance.AppConfig?.ApiEndpoint)) return RemoteStateService.Instance.AppConfig.ApiEndpoint;
            return null;
        }
    }

    /// <summary>
    /// Heartbeat interval override in seconds.
    /// Priority: MDM > remote-state (API) > default (null)
    /// </summary>
    public int? HeartbeatInterval
    {
        get
        {
            if (_isManagedDevice && _heartbeatInterval.HasValue && _heartbeatInterval.Value > 0) return _heartbeatInterval;
            if (RemoteStateService.Instance.AppConfig?.HeartbeatInterval is > 0) return RemoteStateService.Instance.AppConfig.HeartbeatInterval;
            return null;
        }
    }

    /// <summary>
    /// Whether the app should skip all setup UI and go directly to Connected phase.
    /// Requires ManagedSetupComplete=true AND a valid device token.
    /// </summary>
    public bool ShouldSkipAllSetup => ManagedSetupComplete && !string.IsNullOrEmpty(ManagedDeviceToken);

    /// <summary>
    /// Whether the app should skip enrollment UI (but still show setup for cert/proxy).
    /// </summary>
    public bool ShouldSkipEnrollment => (ManagedEnrollmentComplete || ManagedSetupComplete) && !string.IsNullOrEmpty(ManagedDeviceToken);

    private MDMConfigService()
    {
        CheckManagedStatus();
    }

    /// <summary>
    /// Check if this device has MDM configuration and load values.
    /// Called at startup and can be called to refresh.
    /// </summary>
    public void CheckManagedStatus()
    {
        try
        {
            using var key = Registry.LocalMachine.OpenSubKey(PolicyKeyPath);

            if (key == null)
            {
                // No MDM configuration present
                IsManagedDevice = false;
                ClearCachedValues();
                Debug.WriteLine("[MDMConfigService] No MDM configuration found");
                LastConfigCheck = DateTime.UtcNow;
                return;
            }

            // MDM configuration exists - load all values
            IsManagedDevice = true;

            _managedDeviceToken = ReadStringValue(key, KeyDeviceToken);
            _managedDeviceId = ReadStringValue(key, KeyDeviceId);
            _managedWorkspaceId = ReadStringValue(key, KeyWorkspaceId);
            _managedWorkspaceName = ReadStringValue(key, KeyWorkspaceName);
            _managedSetupComplete = ReadBoolValue(key, KeySetupComplete);
            _managedEnrollmentComplete = ReadBoolValue(key, KeyEnrollmentComplete);
            _managedCACertInstalled = ReadBoolValue(key, KeyCACertInstalled);
            _forceAutoStart = ReadBoolValue(key, KeyForceAutoStart);
            _disableUserLogout = ReadBoolValue(key, KeyDisableUserLogout);
            _disableQuit = ReadBoolValue(key, KeyDisableQuit);
            _apiEndpoint = ReadStringValue(key, KeyAPIEndpoint);
            _heartbeatInterval = ReadIntValue(key, KeyHeartbeatInterval);

            LastConfigCheck = DateTime.UtcNow;

            Debug.WriteLine($"[MDMConfigService] MDM configuration loaded:");
            Debug.WriteLine($"  - ManagedDeviceToken: {(string.IsNullOrEmpty(_managedDeviceToken) ? "(not set)" : "(set)")}");
            Debug.WriteLine($"  - ManagedDeviceId: {_managedDeviceId ?? "(not set)"}");
            Debug.WriteLine($"  - ManagedWorkspaceName: {_managedWorkspaceName ?? "(not set)"}");
            Debug.WriteLine($"  - ManagedSetupComplete: {_managedSetupComplete}");
            Debug.WriteLine($"  - ManagedEnrollmentComplete: {_managedEnrollmentComplete}");
            Debug.WriteLine($"  - ForceAutoStart: {_forceAutoStart}");
            Debug.WriteLine($"  - DisableUserLogout: {_disableUserLogout}");
            Debug.WriteLine($"  - DisableQuit: {_disableQuit}");
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"[MDMConfigService] Error reading MDM configuration: {ex.Message}");
            IsManagedDevice = false;
            ClearCachedValues();
        }
    }

    /// <summary>
    /// Apply MDM configuration to the app's standard settings.
    /// Called after CheckManagedStatus() when IsManagedDevice is true.
    /// </summary>
    public void ApplyManagedConfiguration()
    {
        if (!IsManagedDevice)
        {
            Debug.WriteLine("[MDMConfigService] Cannot apply - not a managed device");
            return;
        }

        var settings = Properties.Settings.Default;

        // Apply workspace info if provided
        if (!string.IsNullOrEmpty(_managedWorkspaceName))
        {
            settings.WorkspaceName = _managedWorkspaceName;
            Debug.WriteLine($"[MDMConfigService] Applied WorkspaceName: {_managedWorkspaceName}");
        }

        if (!string.IsNullOrEmpty(_managedWorkspaceId))
        {
            settings.WorkspaceId = _managedWorkspaceId;
            Debug.WriteLine($"[MDMConfigService] Applied WorkspaceId: {_managedWorkspaceId}");
        }

        if (!string.IsNullOrEmpty(_managedDeviceId))
        {
            settings.DeviceId = _managedDeviceId;
            Debug.WriteLine($"[MDMConfigService] Applied DeviceId: {_managedDeviceId}");
        }

        // Apply device token (also triggers token file write in AppState)
        if (!string.IsNullOrEmpty(_managedDeviceToken))
        {
            settings.DeviceToken = _managedDeviceToken;
            Debug.WriteLine("[MDMConfigService] Applied DeviceToken");
        }

        // Mark setup complete if MDM says so
        if (_managedSetupComplete)
        {
            settings.SetupComplete = true;
            Debug.WriteLine("[MDMConfigService] Marked SetupComplete=true");
        }

        settings.Save();
        Debug.WriteLine("[MDMConfigService] Settings saved");
    }

    /// <summary>
    /// Read a string value from registry.
    /// </summary>
    private static string? ReadStringValue(RegistryKey key, string valueName)
    {
        try
        {
            var value = key.GetValue(valueName);
            return value as string;
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Read a boolean value from registry (stored as DWORD, 1=true, 0=false).
    /// </summary>
    private static bool ReadBoolValue(RegistryKey key, string valueName)
    {
        try
        {
            var value = key.GetValue(valueName);
            if (value is int intValue)
                return intValue == 1;
            return false;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Read an integer value from registry.
    /// </summary>
    private static int? ReadIntValue(RegistryKey key, string valueName)
    {
        try
        {
            var value = key.GetValue(valueName);
            if (value is int intValue)
                return intValue;
            return null;
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Clear all cached values when MDM is not present.
    /// </summary>
    private void ClearCachedValues()
    {
        _managedDeviceToken = null;
        _managedDeviceId = null;
        _managedWorkspaceId = null;
        _managedWorkspaceName = null;
        _managedSetupComplete = false;
        _managedEnrollmentComplete = false;
        _managedCACertInstalled = false;
        _forceAutoStart = false;
        _disableUserLogout = false;
        _disableQuit = false;
        _apiEndpoint = null;
        _heartbeatInterval = null;
    }

    #region INotifyPropertyChanged

    public event PropertyChangedEventHandler? PropertyChanged;

    protected bool SetProperty<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
            return false;

        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        return true;
    }

    #endregion
}
